### ISSUE TITLE
splash not setting proper proxy authentication for requests

### ISSUE THREAD

**Issue by pawelmhm** [scrapinghub/splash#344]
It seems like Splash is not setting proper "authentication" headers for requests. I have following Lua script

```lua
function main (splash)
    local host = 'proxy_url'
    local port = 8010
    local user = splash.args.proxy_user
    local password = splash.args.proxy_user
    splash:on_request(function (request)
        request:set_proxy{host, port, username=user, password=password}
    end)
    splash:go(splash.args.url)
    return splash.har()
end
```

when I submit this script to Splash I get 407 from my proxy even though proxy params are correct. When debugging this I see that har log does not have basic_auth header in request headers, and this is the reason why my proxy responds with 407. 

There was related issue #302 but mostly about POST requests.

**Comment by pawelmhm** [comment]
after doing some more research I found out that

1) order of statements in Lua matters:

```lua
    splash:on_request(function (request)
        request:set_proxy{host, port, username=user, password=password}
    end)
    splash:go(splash.args.url)
``` 

fails, but

```lua
splash:go(splash.args.url)    
splash:on_request(function (request)
        request:set_proxy{host, port, username=user, password=password}
    end)
``` 
works.

2) QT Network Access Manager emits signal if proxy rejects credentials: http://doc.qt.io/qt-4.8/qnetworkaccessmanager.html#proxyAuthenticationRequired 

3) QT Network Access Manager caches credentials.

> QNetworkAccessManager will cache the credentials internally. The next time the proxy requests authentication, QNetworkAccessManager will automatically send the same credential without emitting the proxyAuthenticationRequired signal again.

I'm not yet sure what it means for situations when there are multiple requests in splash using different proxies? How this cache is stored? Is it is stored per proxy host? What if there is same proxy host and but it has different usernames and passwords?

From my experience it seems that after first request accepted by proxy others are accepted too. This could mean that network access manager is using cached credentials for further requests with same proxy without checking them. 

I wonder if setting proxy per request is actually good idea, perhaps we should consider moving setProxy to splash object instance instead of to on_request callback? docs say

> setProxy() Sets the proxy to be used in future requests to be proxy. This does not affect requests that have already been sent. The proxyAuthenticationRequired() signal will be emitted if the proxy requests authentication.

so it seems that setProxy should be set per network access manager and not per request.

**Comment by Youwotma** [comment]
I had this problem too, solved it by setting the 'Proxy-Authentication' header manually

**Comment by pawelmhm** [comment]
yeah this is what I'm considering but I wonder if we should not refactor current implementation of setProxy()?

Also small edit on point 1) order of statements - my comment is actually wrong, 2 doesn't work it doesnt set the proxy, I just get 200 response from website without proxying them.

I would definitely vote for adding option to set proxy per splash.

**Comment by Youwotma** [comment]
I think using different proxies for different requests could be an useful feature, if we can get it to work (note that this doesn't exclude adding another method to set the default proxy for every request, similar to the `proxy` parameter in the `render.*` endpoints.

Not sure why this happens, we set the proxy username and password in the QNetworkProxy object when we create it, but maybe the QNetworkAccessManager emits proxyAuthenticationRequired anyway and is expecting us to just do `authenticator.setPassword(proxy.password())`.

I'm going to test this.

**Comment by Youwotma** [comment]
I didn't find why it doesn't work, the `proxyAuthenticationRequired` signal is not emitted no matter what.

**Comment by pawelmhm** [comment]
> I didn't find why it doesn't work, the proxyAuthenticationRequired signal is not emitted no matter what.

I see it is emitted ok, but it looks like this signal is only emitted once. Perhaps in your case you dont see signal being emitted because you already made successful requests to proxy and network access manager cached your credentials? Code referenced above works ok for me but signal is only emitted once.

**Comment by Youwotma** [comment]
Right, I wasn't sending a `Proxy-Authenticate` header with the 407 response and that's why it wasn't emitting the signal.

With your pull request splash enters a infinite loop when the proxy doesn't accept the credentials, maybe we should consider another route. What about generating and setting the `proxy-authorization` header manually in `createRequest` and ignoring the `proxyAuthenticationRequired` signal? That way we save a round trip by not sending the first request that we know it's going to end with 407 and we don't enter an infinite loop if the credentials are invalid (a http407 error is raised in lua).

**Comment by pawelmhm** [comment]
> With your pull request splash enters a infinite loop when the proxy doesn't accept the credentials, maybe we should consider another route

ah yes you're right this should be somehow handled.

> What about generating and setting the proxy-authorization header manually in createRequest and ignoring the proxyAuthenticationRequired signal? T

we could go this way as a workaround but this way we dump QT way of handling proxies. I wonder what's the point of QNetworkProxy and all QT API for proxies if they are useless here? If QNetworkProxy accepts user and password as argument it should also add authorization, either we're doing something wrong here or there's something wrong with API.

I also wonder if Qt api actually properly supports and works well with setting and creating proxies for each request. Would be good to see how things work if you just set proxy per network access manager.

**Comment by Youwotma** [comment]
The QT API is very oriented to implementing a normal non-headless browser, the rationale behind the API is that when the proxyAuthenticationRequired signal is emitted you prompt the user in a blocking modal for the API credentials, that way infinite loops are not a problem. Support for specifying different proxies for different requests in QT also seems poor.

**Comment by pawelmhm** [comment]
> Support for specifying different proxies for different requests in QT also seems poor.

yeah and also this is not exactly common need. I mean how often you need different proxies for different requests? Usually you need one proxy for all requests in one session with some start_url, in this case setting one proxy per splash session would be good thing. But on the other hand it could be difficult to implement because usually (if I'm not mistaken?) there is only one network access manager per Splash and different requests passing through it. So probably this is the reason why we have proxies set per request.

**Comment by kmike** [comment]
Fixed by #348.
